### PR TITLE
Lower memory usage significantly when handling fields/grids

### PR DESCRIPTION
**Pull request by JHolba**
**Issue**
A very large amount of memory is used for loading grids/fields
#5164 

**Approach**
Used memray to pinpoint the issue.
Removed grid that was kept in field config node.
Replaced libecl and xtgeo with ecl_data_io for reading grids.

## Pre review checklist

- [x] Added appropriate release note label
- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Updated documentation
- [x] Ensured new behaviour is tested

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

File: src/ert/_c_wrappers/enkf/config/field_config.py [hunk 1]
```diff
@@ -90,4 +68,6 @@ def save(self, run_path: Path, real_nr: int, ensemble: EnsembleReader):
 
 
 def field_transform(data: npt.ArrayLike, transform_name: str) -> npt.ArrayLike:
+    if transform_name is None:
```
Comment (oyvindeide) [inline src/ert/_c_wrappers/enkf/config/field_config.py:71]: This is called once as:
```
data_transformed = field_transform(data, trans) if trans else data
```
and once as:
```
data_transformed = field_transform(data, trans)
```
suggest converging on one. I personally prefer checking before calling the function as I find that more readable, but that is personal preference.

File: src/ert/storage/field_utils.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+from io import BytesIO
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, List, Optional, TextIO, Tuple, Union
+
+import ecl_data_io
+import numpy as np
+import roffio  # type: ignore
+from xtgeo.grid3d._gridprop_import_roff import import_roff
+
+from ert.storage.grdecl_io import export_grdecl, import_bgrdecl, read_grdecl_3d_property
```
Comment (oyvindeide) [inline src/ert/storage/field_utils.py:16]: Some extra code is introduced here which more or less reproduces functionality that exists in `xtgeo`, I am not convinced it is worth the extra maintenance and code duplication. What is the main reason for doing so? Is it a lot faster, or uses significantly less memory?

I see the argument for `GRID`, as we are just interested in a very small subset, and it is quite memory and computationally heavy to load it through xtgeo, I am however not yet convinced that it is worth re-implementing grid properties, compared to for example suggesting improvements to the api in xtgeo.

**Review by oyvindeide — COMMENTED** [review]
In general this looks like a really good improvement! I am however not convinced we should re-implement reading and writing grid properties, and if we do I think we should increase the test coverage.

File: src/ert/_c_wrappers/enkf/config/field_config.py (hunk 1 above)
Comment (dafeda) [reply src/ert/_c_wrappers/enkf/config/field_config.py:71]: Also, the function `field_transform` states that `transform_name` is a `str` and not an `Optional`.
`field_config.py` is however not checked by `mypy`...

File: src/ert/_c_wrappers/enkf/ensemble_config.py [hunk 3]
```diff
@@ -441,9 +428,9 @@ def get_field_node(
             max_ = float(max_)
         return Field(
             name=name,
-            nx=grid.nx,
-            ny=grid.ny,
-            nz=grid.nz,
+            nx=int(dimensions[0]),
```
Comment (dafeda) [inline src/ert/_c_wrappers/enkf/ensemble_config.py:431]: Have you considered using a `NamedTuple` here instead of a `Tuple`?
We could then for example write `dimension.nx` instead of `dimension[0]` which I think reads better.

File: src/ert/storage/local_ensemble.py [hunk 4]
```diff
@@ -357,15 +353,24 @@ def load_gen_data_as_df(
         return pd.concat(dfs)
 
     def load_field(self, key: str, realizations: List[int]) -> npt.NDArray[np.double]:
-        result = []
+        result: Optional[npt.NDArray[np.double]] = None
         for realization in realizations:
             input_path = self.mount_point / f"realization-{realization}"
             if not input_path.exists():
                 raise KeyError(f"Unable to load FIELD for key: {key}")
```
Comment (dafeda) [inline src/ert/storage/local_ensemble.py:360]: Would it be better if the error message also mentioned the realization-number?
Something like this perhaps:

f"Unable to load FIELD for key {key} in realization {realization}."

File: src/ert/storage/local_ensemble.py [hunk 5]
```diff
@@ -357,15 +353,24 @@ def load_gen_data_as_df(
...
             data = np.load(input_path / f"{key}.npy")
-            data_no_nans = data[~np.isnan(data)]
-            result.append(data_no_nans)
-        return np.stack(result).T  # type: ignore
+            data = data[np.isfinite(data)]
+
+            if result is None:
```
Comment (dafeda) [inline src/ert/storage/local_ensemble.py:364]: Are we sure that all the fields produced will have the same elements be `NaN` or `inf`?
Perhaps add a check like the following?

```python
            if result is None:
                result = np.empty((len(realizations), data.size), dtype=np.double)
            elif data.size != result.shape[1]:
                raise ValueError(f"Data size mismatch for realization {realization}")
```

File: src/ert/storage/local_ensemble.py [hunk 6]
```diff
@@ -357,15 +353,24 @@ def load_gen_data_as_df(
...
+                result = np.ndarray((len(realizations), data.size))
```
Comment (dafeda) [inline src/ert/storage/local_ensemble.py:365]: What if we instead initialize with this:

`result = np.empty((data.size, len(realizations)), dtype=np.double)

and then do

`result[:, realization] = data`

to avoid having to transpose before returning.

Transpose is cheap, but this is still more explicit I think.

File: src/ert/storage/local_ensemble.py [hunk 7]
```diff
@@ -357,15 +353,24 @@ def load_gen_data_as_df(
...
+        result: Optional[npt.NDArray[np.double]] = None
         for realization in realizations:
             input_path = self.mount_point / f"realization-{realization}"
             if not input_path.exists():
                 raise KeyError(f"Unable to load FIELD for key: {key}")
             data = np.load(input_path / f"{key}.npy")
-            data_no_nans = data[~np.isnan(data)]
-            result.append(data_no_nans)
-        return np.stack(result).T  # type: ignore
+            data = data[np.isfinite(data)]
```
Comment (dafeda) [inline src/ert/storage/local_ensemble.py:362]: What if we try using memory-mapping here?
So, I have not tried this but the docs of `np.load` state that:

_mmap_mode{None, ‘r+’, ‘r’, ‘w+’, ‘c’}, optional
If not None, then memory-map the file, using the given mode (see [numpy.memmap](https://numpy.org/doc/stable/reference/generated/numpy.memmap.html#numpy.memmap) for a detailed description of the modes). A memory-mapped array is kept on disk. However, it can be accessed and sliced like any ndarray. Memory mapping is especially useful for accessing small fragments of large files without reading the entire file into memory._

So something like this should work I think:

```python
# Memory-map the .npy file
data = np.load(input_path / f"{key}.npy", mmap_mode='r')
# Create a boolean mask for finite values
finite_mask = np.isfinite(data)
# Load only finite values
finite_data = data[finite_mask]
```

File: src/ert/storage/local_ensemble.py [hunk 8]
```diff
@@ -655,7 +595,7 @@ def save_gen_kw(  # pylint: disable=R0913
         parameter_name: str,
         parameter_keys: List[str],
         realizations: List[int],
-        data: npt.NDArray[np.float64],
+        data: npt.NDArray[np.double],
```
Comment (dafeda) [inline src/ert/storage/local_ensemble.py:598]: Not really part of this PR, but you can safely remove "# pylint: disable=R0913" next to `def save_gen_kw` as the function does not have "too-many-arguments".

File: src/ert/_c_wrappers/enkf/config/field_config.py (hunk 1 above)
Comment (JHolba) [reply src/ert/_c_wrappers/enkf/config/field_config.py:71]: will change transform_name to optional and check inside field_transform. I kind of agree with @oyvindeide about checking before calling, but since there is no enforcement of types in Python you will still need to check inside that function to prevent exceptions.

File: src/ert/storage/local_ensemble.py (hunk 8 above)
Comment (JHolba) [reply src/ert/storage/local_ensemble.py:600]: removed it

File: src/ert/_c_wrappers/enkf/ensemble_config.py (hunk 3 above)
Comment (JHolba) [reply src/ert/_c_wrappers/enkf/ensemble_config.py]: added a NamedTuple named Shape

File: src/ert/storage/local_ensemble.py (hunk 6 above)
Comment (JHolba) [reply src/ert/storage/local_ensemble.py:352]: I only did it to keep it the same as the original. The transpose may return a view instead of a copy and depending on how the array is accessed later, that may be beneficial. If you think it's better to create it this way, I will change it.

File: src/ert/storage/local_ensemble.py (hunk 7 above)
Comment (JHolba) [reply src/ert/storage/local_ensemble.py:349]: This was added

File: src/ert/storage/local_ensemble.py (hunk 6 above)
Comment (dafeda) [reply src/ert/storage/local_ensemble.py:352]: Let's leave it as is for now.

File: src/ert/storage/field_utils.py (hunk 2 above)
Comment (dafeda) [reply src/ert/storage/field_utils.py]: Where are we one this one?
I also tend to believe that it's better to extend and improve `xtgeo`, but I have not looked at the design of `xtgeo` and perhaps it's too difficult to extend it at this point.

File: src/ert/storage/local_ensemble.py (hunk 5 above)
Comment (JHolba) [reply src/ert/storage/local_ensemble.py:351]: we are sure that they will always have the same cells masked with NaNs as the code is now, but it doesn't hurt to add that test.

File: src/ert/storage/field_utils.py (hunk 2 above)
Comment (JHolba) [reply src/ert/storage/field_utils.py]: I think we should merge it as is just to get this done. We can make a new issue to go back to using xtgeo for field loading if we wish.

Comment (dafeda) [reply src/ert/storage/field_utils.py]: I am fine by that Jon.

**Review by dafeda — APPROVED** [review]
This is stellar work 👍
